docs/content/guides/dagster/ml-ops.mdx

```diff
@@ -0,0 +1,221 @@
+from sklearn.feature_extraction.text import TfidfVectorizer
+import numpy as np
+
+@multi_asset(outs={'Tfidf_Vectorizer': AssetOut(), 'transformed_training_data': AssetOut()})
+def transformed_train(training_data):
+    X_train,  y_train = training_data
+    """Initiate and fit the tokenizer on the training data and transform the training dataset"""
+    vectorizer = TfidfVectorizer()
+    transformed_X_train = vectorizer.fit_transform(X_train)
+    transformed_X_train = transformed_X_train.toarray()
+    y_train = y_train.fillna(0)
+    transformed_y_train = np.array(y_train)
+    return vectorizer, (transformed_X_train, transformed_y_train)
+
+@asset
+def transformed_test_data(test_data, Tfidf_Vectorizer):
+    X_test, y_test = test_data
+    """Use the fitted tokenizer to transform the test dataset"""
+    transformed_X_test = Tfidf_Vectorizer.transform(X_test)
+    transformed_y_test = np.array(y_test)
+    y_test = y_test.fillna(0)
+    transformed_y_test = np.array(y_test)
+    return transformed_X_test, transformed_y_test
+```
+
+We have also transformed the dataframes into numpy arrays and removed na values to prepare the data for training.
+
+### Model training
+
+At this point, we have X_train, y_train, X_test, y_test ready to go for our model. We can use any number of models to train our model from libraries like [sklearn](https://scikit-learn.org/), [tensorflow](https://www.tensorflow.org/) and [pytorch](https://pytorch.org/).
+
+In our example, we will be training a [XGBoost model](https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRegressor) to predict a numerical value.
+
+```python file=/guides/dagster/ml_ops/ml_ops.py startafter=models_start endbefore=models_end
+import xgboost as xg
+from sklearn.metrics import mean_absolute_error
+
+@asset
+def xgboost(transformed_training_data):
+    transformed_X_train, transformed_y_train = transformed_training_data
+    """Train XGBoost model, which is a highly efficent and flexible model"""
+    xgb_r = xg.XGBRegressor(objective ='reg:squarederror', eval_metric=mean_absolute_error,
+                  n_estimators = 20)
+    xgb_r.fit(transformed_X_train, transformed_y_train)
+    return xgb_r
+
+@asset
+def score_xgboost( transformed_test_data, xgboost):
+    transformed_X_test, transformed_y_test = transformed_test_data
+    """Use the test set data to get a score of the XGBoost model"""
+    score = xgboost.score(transformed_X_test, transformed_y_test)
+    return score
+```
+
+We include freshness policies in the models here which will ensure the upstream data transformation assets are updated to comply with these policies.
+
+### Evaluate our results
+
+In our model assets, we evaluated each of the models on the test data and in this case, got the error between the predicted and actual results. Next, lets create an asset which runs inference on the model, more frequently than the model is re-trained to predict the results.
```

Comment:
> In our model assets, we evaluated each of the models on the test data and in this case, got the error between the predicted and actual results.

What error? I wasn't instructed to run or look at anything, so how can I know there's an error?

> Next, lets create an asset which runs inference on the model, more frequently than the model is re-trained to predict the results.

I'm finding this hard to follow - I think it's because of how it's structured. Is this accurate?:

> Next, to predict the results, we'll create another asset that runs inference on the model more frequently than the model is re-trained.

Reply:
updated- do we know how to put R^2 in nicely?